### PR TITLE
CORS

### PR DESCRIPTION
**Pull request by nitinmohan87**

### CODE REVIEW

File: cors/Makefile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# Targets:
+# - "gen" generates the goa files for the example services
+
+PLUGIN_NAME=cors
+ALIASER_SRC=goa.design/goa/http/dsl
+
+# include common Makefile content for plugins
+include $(GOPATH)/src/goa.design/plugins/plugins.mk
+
+gen:
```
Comment (raphael) [inline cors/Makefile]: I guess this needs to be updated?

File: cors/design/aliases.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
+//************************************************************************//
```
Comment (raphael) [inline cors/design/aliases.go]: Do we need this file?

File: cors/design/cors.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+			}
+		}
+		if !found {
+			paths = append(paths, fp)
+		}
+	}
+	return paths
+}
+
+// EvalName returns the generic definition name used in error messages.
```
Comment (raphael) [inline cors/design/cors.go]: definition -> expression (in all comments)

File: cors/design/root.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	oexps = make(eval.ExpressionSet, 0, len(r.ServiceOrigins))
+	for _, o := range r.ServiceOrigins {
+		oexps = append(oexps, o)
+	}
+	walk(oexps)
+}
+
+// DependsOn tells the eval engine to run the goa DSL first.
+func (r *RootExpr) DependsOn() []eval.Root {
+	return []eval.Root{goadesign.Root}
```
Comment (raphael) [inline cors/design/root.go]: I think this needs to be the `http` root?

File: cors/design/root.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+// Packages returns the import path to the Go packages that make
+// up the DSL. This is used to skip frames that point to files
+// in these packages when computing the location of errors.
+func (r *RootExpr) Packages() []string {
+	return []string{}
```
Comment (raphael) [inline cors/design/root.go]: Shouldn't this be `[]string{"goa.design/plugins/cors/dsl"}`?

File: cors/dsl/cors.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+	goadesign "goa.design/goa/design"
+	"goa.design/goa/eval"
+	"goa.design/plugins/cors/design"
+)
+
+// Origin defines the CORS policy for a given origin. The origin can use a wildcard prefix
+// such as "https://*.mydomain.com". The special value "*" defines the policy for all origins
+// (in which case there should be only one Origin DSL in the parent resource).
+// The origin can also be a regular expression wrapped into "/".
+// Example:
```
Comment (raphael) [inline cors/dsl/cors.go]: Remove `Example:`?

File: cors/dsl/cors.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+//
+// Origin must appear in API or Service Expression.
+//
+// Origin accepts an origin string as the first argument and
+// an optional DSL function as the second argument.
+//
+// Example:
+//
+//		var _ = API("calc", func() {
+//        Origin("http://swagger.goa.design", func() { // Define CORS policy, may be prefixed with "*" wildcard
```
Comment (raphael) [inline cors/dsl/cors.go]: Fix indenting

File: cors/dsl/cors.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+	goadesign "goa.design/goa/design"
+	"goa.design/goa/eval"
+	"goa.design/plugins/cors/design"
+)
+
+// Origin defines the CORS policy for a given origin. The origin can use a wildcard prefix
+// such as "https://*.mydomain.com". The special value "*" defines the policy for all origins
+// (in which case there should be only one Origin DSL in the parent resource).
+// The origin can also be a regular expression wrapped into "/".
```
Comment (raphael) [inline cors/dsl/cors.go]: >The origin can also be a regular expression in which case it must be wrapped with "/".

File: cors/dsl/cors.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+//        })
+//		})
+//
+//		var _ = Service("calculator", func() {
+//        Origin("/(api|swagger)[.]goa[.]design/") // Define CORS policy with a regular expression
+//
+//        Method("add", func() {
+//            Description("Add two operands")
+//
+//            // Override default service security requirements. Require
```
Comment (raphael) [inline cors/dsl/cors.go]: invalid (security) example

File: cors/dsl/cors.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+// Methods sets the origin allowed methods.
+//
+// Methods must be used in an Origin expression.
+//
+// Example:
+//
+// Origin("http://swagger.goa.design", func() {
+//				 Methods("GET", "POST")								// One or more authorized HTTP methods
```
Comment (raphael) [inline cors/dsl/cors.go]: Fix indenting

File: cors/dsl/cors.go [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+	}
+}
+
+// Headers sets the authorized headers. "*" authorizes all headers.
+//
+// Headers must be used in an Origin expression.
+//
+// Example:
+//
+// Origin("http://swagger.goa.design", func() {
```
Comment (raphael) [inline cors/dsl/cors.go]: Fix indenting (code example should be indented)
Applies to all comments below.

File: cors/examples/cellar/design/sommelier.go [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+	. "goa.design/goa/http/design"
+	. "goa.design/plugins/cors/dsl"
+)
+
+var _ = Service("sommelier", func() {
+	Description("The sommelier service retrieves bottles given a set of criteria.")
+	HTTP(func() {
+		Path("/sommelier")
+	})
+	Origin("/.*localhost*/", func() {
```
Comment (raphael) [inline cors/examples/cellar/design/sommelier.go]: missing `.` before second `*`

File: cors/Makefile (hunk 1 above)
Comment (nitinmohan87) [reply cors/Makefile]: Done

File: cors/dsl/cors.go (hunk 11 above)
Comment (nitinmohan87) [reply cors/dsl/cors.go]: Done

**Pull request merged**